### PR TITLE
renderer_vulkan: Add vulkan initialization code

### PR DESCRIPTION
**Pull request by GPUCode**
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6620)

### CODE REVIEW

File: src/common/CMakeLists.txt [hunk 1]
```diff
@@ -65,8 +65,8 @@ add_library(citra_common STATIC
     common_precompiled_headers.h
     common_types.h
     construct.h
-    dynamic_library/dynamic_library.cpp
-    dynamic_library/dynamic_library.h
```
Comment (Steveice10) [inline src/common/CMakeLists.txt:69]: Not sure why this needs to be moved, before it was kept together with the other dynamic library code.

File: src/video_core/renderer_vulkan/pica_to_vk.h [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+inline vk::Filter TextureFilterMode(TextureFilter mode) {
+    switch (mode) {
+    case TextureFilter::Linear:
+        return vk::Filter::eLinear;
+    case TextureFilter::Nearest:
+        return vk::Filter::eNearest;
+    default:
+        LOG_CRITICAL(Render_Vulkan, "Unknown texture filtering mode {}", mode);
+        UNIMPLEMENTED();
```
Comment (SachinVin) [inline src/video_core/renderer_vulkan/pica_to_vk.h:24]: Can probably use `UNIMPLEMENTED_MSG`

File: src/video_core/renderer_vulkan/pica_to_vk.h [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        vk::SamplerAddressMode::eClampToEdge,
+        vk::SamplerAddressMode::eClampToBorder,
+        vk::SamplerAddressMode::eRepeat,
+        vk::SamplerAddressMode::eRepeat,
+    }};
+
+    const auto index = static_cast<std::size_t>(mode);
+
+    // Range check table for input
+    if (index >= wrap_mode_table.size()) {
```
Comment (SachinVin) [inline src/video_core/renderer_vulkan/pica_to_vk.h:61]: can be simplified to assert ?

File: src/video_core/renderer_vulkan/pica_to_vk.h (hunk 2 above)
Comment (GPUCode) [reply src/video_core/renderer_vulkan/pica_to_vk.h:24]: unimplemented denotes something that should be implemented, but here all the enums known are handled

Comment (SachinVin) [reply src/video_core/renderer_vulkan/pica_to_vk.h:24]: Either you, or I have misunderstood something

Comment (GPUCode) [reply src/video_core/renderer_vulkan/pica_to_vk.h:24]: Probably me lol. In this example TextureFilter only has Linear and Nearest enums which are both handled in the switch, so technically there isn't something that's "unimplemented" (I know, this is kinda approaching nitpick territory). It's not too big a deal to add this though, just a thought

Comment (SachinVin) [reply src/video_core/renderer_vulkan/pica_to_vk.h:24]: So it should be changed to unreachable?

Comment (SachinVin) [reply src/video_core/renderer_vulkan/pica_to_vk.h:24]: So it should be changed to unreachable?

File: src/video_core/renderer_vulkan/pica_to_vk.h [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        return vk::Filter::eNearest;
+    default:
+        LOG_CRITICAL(Render_Vulkan, "Unknown texture filtering mode {}", mode);
+        UNIMPLEMENTED();
+    }
+
+    return vk::Filter::eLinear;
+}
+
+inline vk::SamplerMipmapMode TextureMipFilterMode(TextureFilter mip) {
```
Comment (FearlessTobi) [inline src/video_core/renderer_vulkan/pica_to_vk.h:30]: is inline needed here?

File: src/video_core/renderer_vulkan/vk_instance.h [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+        return present_queue;
+    }
+
+    /// Returns true if logic operations need shader emulation
+    bool NeedsLogicOpEmulation() const {
+        return !features.logicOp;
+    }
+
+    bool UseGeometryShaders() const {
+#ifndef __ANDROID__
```
Comment (FearlessTobi) [inline src/video_core/renderer_vulkan/vk_instance.h:99]:
```suggestion
#ifdef __ANDROID__
```

Let's handle the positive case first

File: src/video_core/renderer_vulkan/vk_instance.cpp [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+
+    const bool is_arm = driver_id == vk::DriverIdKHR::eArmProprietary;
+    const bool is_qualcomm = driver_id == vk::DriverIdKHR::eQualcommProprietary;
+
+    add_extension(VK_KHR_SWAPCHAIN_EXTENSION_NAME);
+    image_format_list = add_extension(VK_KHR_IMAGE_FORMAT_LIST_EXTENSION_NAME);
+    shader_stencil_export = add_extension(VK_EXT_SHADER_STENCIL_EXPORT_EXTENSION_NAME);
+    bool has_timeline_semaphores = add_extension(VK_KHR_TIMELINE_SEMAPHORE_EXTENSION_NAME,
+                                                 is_qualcomm, "it is broken on Qualcomm drivers");
+    bool has_portability_subset = add_extension(VK_KHR_PORTABILITY_SUBSET_EXTENSION_NAME);
```
Comment (FearlessTobi) [inline src/video_core/renderer_vulkan/vk_instance.cpp:408]: missing a few `const`'s here

**Review by FearlessTobi — APPROVED** [review]

File: src/video_core/renderer_vulkan/pica_to_vk.h (hunk 4 above)
Comment (GPUCode) [reply src/video_core/renderer_vulkan/pica_to_vk.h:30]: Yes it's required to define functions in the header file

**Review by Steveice10 — APPROVED** [review]

**Comment by GPUCode** [comment]
Merging to unblock vulkan PR

**Pull request merged**